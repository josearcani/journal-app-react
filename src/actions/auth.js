import { getAuth, signInWithPopup } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

// async action
export const startLoginEmailPassword = (email, password) => {
  // use fetch or axios to get data

  return (dispatch) => {
    setTimeout(() => {
      dispatch( login('asfasi23534on32', 'Gerardo') )
    }, 1500);
  }
}

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
      })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})
