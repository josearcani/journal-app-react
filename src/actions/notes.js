import { addDoc, collection } from "firebase/firestore";
import { db } from "../firebase/firebase-config";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    // { auth, notes, ui }
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    }
    
    const docRef = await addDoc(collection(db ,`${uid}/journal/notes`), newNote);

    console.log(docRef);
  }
}