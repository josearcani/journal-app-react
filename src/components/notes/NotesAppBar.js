import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

  const dispatch = useDispatch();

  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote(active));
  }

  const handleUploadPicture = () => {
    document.querySelector('#fileSelector').click();
  }

  const handleFileChange = (e) => {
    // e.target => <input />
    // e.target.files ==> FileList {0: File, length: 1}
    const file = e.target.files[0];
    dispatch(startUploading(file));
  }

  return (
    <div className="notes__appbar">
      <span>28 de septiembre 2021</span>

      <input
        id="fileSelector"
        type="file"
        name="file"
        style={{ display: "none" }}
        onChange={ handleFileChange }
      />

      <div>
        <button
          className="btn"
          onClick={ handleUploadPicture }
        >
          Picture
        </button>

        <button
          className="btn"
          onClick={ handleSave }
        >
          Save
        </button>
      </div>
    </div>
  )
}
