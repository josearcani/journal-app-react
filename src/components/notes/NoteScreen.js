import React, { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux';
import { noteActive, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {

  const dispatch = useDispatch();
  const { active: note } = useSelector(state => state.notes);
  const [ formValues, handleInputChange, cleanForm ] = useForm(note);
  
  const activeId = useRef(note.id);
  // this ref is just for the first note

  useEffect(() => {
    if (activeId.current !== note.id) {
      cleanForm(note);
      activeId.current = note.id;
      // update the new id to reference and clean with new active note
    }
  }, [ note, cleanForm ])


  useEffect(() => {
    dispatch(noteActive(formValues.id, { ...formValues }))
  }, [ dispatch ,formValues ])

  const handleDelete = () => {
    dispatch(startDeleting(formValues.id))
  }

  return (
    <div className="notes__main-content">

      <NotesAppBar />

      <div className="notes__content">

        <input
          type="text"
          placeholder="Some delightful thought"
          className="notes__title-input"
          autoComplete="off"
          name="title"
          onChange={ handleInputChange }
          value={ formValues.title }
        />

        <textarea
          placeholder="Hey write over here..."
          className="notes__textarea"
          name="body"
          onChange={ handleInputChange }
          value={ formValues.body }
        ></textarea>


        {
          note.url &&
              (<div className="notes__image">
                <img
                  src={ note.url }
                  alt={ formValues.title }
                />
              </div>)
        }

      </div>

        <button
          className="btn btn-danger"
          onClick={ handleDelete }
        >
          Delete
        </button>

    </div>
  )
}
