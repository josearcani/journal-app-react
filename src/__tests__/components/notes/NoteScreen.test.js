import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { noteActive } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
  noteActive: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth: {
    uid:'1',
    name: 'Joseph'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    active: {
      id: '1234',
      title: 'Hello',
      body: 'world',
      date: 699669
    },
    notes: [],
  }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={ store }>
    <NoteScreen />
  </Provider>
);

describe('Test on <NoteScreen />', () => {

  test('should NoteScreen render normally', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call activeNote action when changing values', () => {
    wrapper.find('input[type="text"]').simulate('change', {
      target: { name: 'title', value: 'secret' }
    })

    expect(noteActive).toHaveBeenCalled(); 
    expect(noteActive).toHaveBeenLastCalledWith('1234', {
      id: '1234',
      title: 'secret',
      body: 'world',
      date: 699669
    });
  });
});
