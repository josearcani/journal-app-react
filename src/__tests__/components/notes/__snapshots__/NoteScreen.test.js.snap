// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <NoteScreen /> should NoteScreen render normally 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de septiembre 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some delightful thought"
      type="text"
      value="Hello"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="Hey write over here..."
      value="world"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
